# Vanessa WIlliams

## Import libraries

### After calling in the libraries needed to run codes, I downnloaded a program within NLTK library to remove common English words, loaded the data to "clean". and displaye all of the dataset.

In [1]:
# Import necessary libraries
import pandas as pd
import re
import string
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import nltk
from nltk.corpus import stopwords

# Download NLTK stopwords if needed
nltk.download('stopwords')
stop_words = set(stopwords.words('english'))

# Load the dataset from your local folder (replace with the correct path)
file_path = '/Users/vanessawilliams/Desktop/Vanessa_Williams/twitter_sample.csv'
df = pd.read_csv(file_path)

# Display the first few rows of the dataset
df.head()

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


FileNotFoundError: [Errno 2] No such file or directory: '/Users/vanessawilliams/Desktop/Vanessa_Williams/twitter_sample.csv'

## Clean the Tweet Content


### Selected the text that I want to clean. To do so, I tokenized the text. This prepares text data for NLP tasks. Then i called a function that removes retweets and display the cleaned data/

In [ ]:
# Clean the "Tweet Content" column by removing non-text data and stopwords
def clean_text(text):
    # RLs, mentioRemove non-text data like Uns, and hashtags
    text = re.sub(r'http\S+|www\S+|https\S+|@\S+|#\S+', '', text, flags=re.MULTILINE)
    # Remove punctuation
    text = text.translate(str.maketrans('', '', string.punctuation))
    # Tokenize text and remove stopwords
    tokens = text.lower().split()
    tokens = [word for word in tokens if word not in stop_words]
    # Join tokens back into a single string
    return ' '.join(tokens)

# Apply the cleaning function to the "Tweet Content" column
df['cleaned_tweet'] = df['Tweet Content'].apply(clean_text)

# Filter only tweets (removing retweets)
df = df[~df['cleaned_tweet'].str.contains('rt', case=False)]

# Display the cleaned data
df.head()

Tweet Id                                          Tweet URL   
2   "1167237977615097861"  https://twitter.com/lordbyronaf/status/1167237...  \
3   "1167236897078480898"  https://twitter.com/CountessDavis/status/11672...   
4   "1167228378191204353"  https://twitter.com/Local12/status/11672283781...   
5   "1167228285463531520"  https://twitter.com/lbonis1/status/11672282854...   
12  "1166892836165496835"  https://twitter.com/AndreaWestbyMD/status/1166...   

   Tweet Posted Time (UTC)                                      Tweet Content   
2             8/30/19 0:49  What a great team ⁦@HealthSourceOH⁩ ⁦@Local12⁩...  \
3             8/30/19 0:45  What a great team ⁦@HealthSourceOH⁩ ⁦@Local12⁩...   
4             8/30/19 0:11  What a great team ⁦@HealthSourceOH⁩ ⁦@Local12⁩...   
5             8/30/19 0:11  What a great team ⁦@HealthSourceOH⁩ ⁦@Local12⁩...   
12            8/29/19 1:58  Y’all ⁦@UMNNMFamMedRes⁩ Rose Marie Leslie is a...   

   Tweet Type               Client  Retweets Received  Likes Received   
2     ReTweet  Twitter for Android                  0               0  \
3     ReTweet  Twitter for Android                  0               0   
4     ReTweet            TweetDeck                  0               0   
5       Tweet   Twitter for iPhone                  3              17   
12      Tweet   Twitter for iPhone                  0              27   

     Tweet Location Tweet Language  ...        Username   
2         Ohio, USA        English  ...     lordbyronaf  \
3               NaN        English  ...   CountessDavis   
4    Cincinnati, OH        English  ...         Local12   
5           WKRC TV        English  ...         lbonis1   
12  Minneapolis, MN        English  ...  AndreaWestbyMD   

                                             User Bio   
2   It's easier to be who you are, than it is to b...  \
3   I am named after @ElvisPresley daughter Lisa M...   
4   Local 12 is #Cincinnati's trusted source for b...   
5   Health and Medical Reporter/News Anchor Regist...   
12  She/her/hers. Teacher of family medicine, form...   

   Verified or Non-Verified                         Profile URL   
2              Non-Verified     https://twitter.com/lordbyronaf  \
3              Non-Verified   https://twitter.com/CountessDavis   
4                  Verified         https://twitter.com/Local12   
5                  Verified         https://twitter.com/lbonis1   
12             Non-Verified  https://twitter.com/AndreaWestbyMD   

   Protected or Non-protected User Followers User Following   
2               Non-Protected           7808           8617  \
3               Non-Protected            291             81   
4               Non-Protected         198675            651   
5               Non-Protected           6015           4866   
12              Non-Protected            938           1247   

    User Account Creation Date  Impressions   
2                7/25/12 15:43            0  \
3                1/26/17 18:21            0   
4                 9/2/08 20:09            0   
5                3/16/13 12:05        12033   
12               4/15/15 19:58         1876   

                                        cleaned_tweet  
2                                      great team ⁦ ⁦  
3                                      great team ⁦ ⁦  
4                                      great team ⁦ ⁦  
5                                      great team ⁦ ⁦  
12  y’all ⁦ rose marie leslie ⁦ blue ribbon winnin...  

[5 rows x 22 columns]

# Remove Duplicate Tweets

In [ ]:
# Drop duplicate tweets based on the 'cleaned_tweet' column
df = df.drop_duplicates(subset='cleaned_tweet', keep='first')

# Verify if duplicates were removed
df['cleaned_tweet'].value_counts()

cleaned_tweet
great team ⁦ ⁦                                                                                                                                          1
2 episodes left july favorite past year haven’t played yet got guess next 2 might days till next one                                                    1
weve loved motivated stories social prescribing initiatives around world continue conversations us national conference november well discussing need    1
doubt incredible job full challenges beyond rewarding weekends like great reminder work hard love                                                       1
agree                                                                                                                                                   1
                                                                                                                                                       ..
vaccines help prevent illness else protect animals ⬇️ heres so

# Bag of Words (BoW) Representation

In [ ]:
# Step 2: Bag of Words (BoW) Representation
cv = CountVectorizer()
cv_matrix = cv.fit_transform(df['cleaned_tweet'])
vocab = cv.get_feature_names_out()

# Convert the BoW matrix to a DataFrame for better visualization
bow_df = pd.DataFrame(cv_matrix.toarray(), columns=vocab)
print("Bag of Words Representation:\n", bow_df.head())

Bag of Words Representation:
    10  10000  12  2010  2019  21  25  40  5000th  600  ...  year  years   
0   0      0   0     0     0   0   0   0       0    0  ...     0      0  \
1   0      0   0     0     0   0   0   0       0    0  ...     0      0   
2   0      0   0     0     0   0   0   0       0    0  ...     0      0   
3   0      0   0     0     0   0   0   0       0    0  ...     0      0   
4   0      0   0     0     0   0   0   0       0    0  ...     0      0   

   yesterday  yet  young  youre  zoonosen  zu  zwierząt  że  
0          0    0      0      0         0   0         0   0  
1          0    0      0      0         0   0         0   0  
2          0    0      0      0         0   0         0   0  
3          0    0      0      0         1   0         0   0  
4          0    0      0      0         0   0         2   1  

[5 rows x 623 columns]


## TF-IDF Representation

### Step 3: TF-IDF Representation
### Convert the TF-IDF matrix to a DataFrame for better visualization
### Display results

In [ ]:
# Step 3: TF-IDF Representation
tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(df['cleaned_tweet'])
tfidf_vocab = tfidf.get_feature_names_out()

# Convert the TF-IDF matrix to a DataFrame for better visualization
tfidf_df = pd.DataFrame(tfidf_matrix.toarray(), columns=tfidf_vocab)
print("\nTF-IDF Representation:\n", tfidf_df.head())


TF-IDF Representation:
     10  10000   12  2010  2019   21   25   40  5000th  600  ...  year  years   
0  0.0    0.0  0.0   0.0   0.0  0.0  0.0  0.0     0.0  0.0  ...   0.0    0.0  \
1  0.0    0.0  0.0   0.0   0.0  0.0  0.0  0.0     0.0  0.0  ...   0.0    0.0   
2  0.0    0.0  0.0   0.0   0.0  0.0  0.0  0.0     0.0  0.0  ...   0.0    0.0   
3  0.0    0.0  0.0   0.0   0.0  0.0  0.0  0.0     0.0  0.0  ...   0.0    0.0   
4  0.0    0.0  0.0   0.0   0.0  0.0  0.0  0.0     0.0  0.0  ...   0.0    0.0   

   yesterday  yet  young  youre  zoonosen   zu  zwierząt        że  
0        0.0  0.0    0.0    0.0  0.000000  0.0  0.000000  0.000000  
1        0.0  0.0    0.0    0.0  0.000000  0.0  0.000000  0.000000  
2        0.0  0.0    0.0    0.0  0.000000  0.0  0.000000  0.000000  
3        0.0  0.0    0.0    0.0  0.353553  0.0  0.000000  0.000000  
4        0.0  0.0    0.0    0.0  0.000000  0.0  0.460832  0.230416  

[5 rows x 623 columns]


# Cosine Similarity

### Step 4: Cosine Similarity between documents (tweets)
### Display Results

In [ ]:
# Step 4: Cosine Similarity between documents (tweets)
cosine_sim = cosine_similarity(tfidf_matrix)
print("\nCosine Similarity Matrix:\n", cosine_sim)


Cosine Similarity Matrix:
 [[1.         0.         0.         ... 0.         0.         0.13185464]
 [0.         1.         0.         ... 0.         0.         0.        ]
 [0.         0.         1.         ... 0.         0.         0.        ]
 ...
 [0.         0.         0.         ... 1.         0.         0.        ]
 [0.         0.         0.         ... 0.         1.         0.21101903]
 [0.13185464 0.         0.         ... 0.         0.21101903 1.        ]]


# Find Most Similar Tweets

### Step 5: Find most similar tweets to the first tweet (index 0)
### Sort the tweets by similarity in descending order
### Display Results


In [ ]:
# Step 5: Find most similar tweets to the first tweet (index 0)
similar_tweets = cosine_sim[0]

# Sort the tweets by similarity in descending order
similar_indices = similar_tweets.argsort()[-5:][::-1]
print("\nTop 5 most similar tweets to the first tweet:")
for idx in similar_indices:
    print(f"Tweet {idx} - Cosine Similarity: {similar_tweets[idx]}")
    print(df['cleaned_tweet'].iloc[idx], "\n")


Top 5 most similar tweets to the first tweet:
Tweet 0 - Cosine Similarity: 1.0
great team ⁦ ⁦ 

Tweet 10 - Cosine Similarity: 0.2588208505551044
delivered 5000th special aflac duck yesterday proud team 

Tweet 77 - Cosine Similarity: 0.17381648047381992
great stats shared french member annual conference 64 new since 2010 

Tweet 58 - Cosine Similarity: 0.1537358270295723
doubt incredible job full challenges beyond rewarding weekends like great reminder work hard love 

Tweet 80 - Cosine Similarity: 0.13185463897514438
great weekend back month july rebroadcasting best past year crew takes much needed break transitions new academic year 



### Summary of Results:

The final output shows the top 5 tweets that are most similar to the first tweet in the dataset, based on **Cosine Similarity** using their **TF-IDF representations**.

- **Tweet 0** has a **Cosine Similarity of 1.0** because it is being compared to itself.
- The other tweets have lower similarity scores, indicating varying levels of similarity to the first tweet:
  - **Tweet 10** has a similarity score of **0.2588**, making it the most similar to the first tweet.
  - **Tweet 77** and **Tweet 58** also show moderate similarity with scores of **0.1738** and **0.1537** respectively.
  - **Tweet 80** has the lowest similarity score in the top 5 at **0.1319**.

### Cosine Similarity Explanation:

I used Cosine Similarity to measure the similarity between tweets. It compares the **TF-IDF vector representations** of each tweet to calculate the similarity based on the angle between their vectorized representations in multi-dimensional space. A score of **1.0** indicates identical content, while scores closer to **0** indicate less similarity.

### Observations:

- **Cleaning and pre-processing**: After cleaning the dataset by removing non-text content and duplicates, we observed a more meaningful distribution of similarity scores.
- **Variation in similarity**: Without duplicates, tweets that share similar content (such as references to teams, events, or general sentiments) appear as more similar, while tweets with different content show lower similarity scores.
- **Insights**: Cosine Similarity was effective in identifying tweets that share similar language and topics, while successfully differentiating between tweets with less content overlap.